### PR TITLE
Add the Plywrks ply8x

### PR DESCRIPTION
**Pull request by ramonimbao**
## Description

Adds the Plywrks ply8x: F12 TKL PCB.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/plywrks/ply8x/rules.mk [hunk 1]
```diff
@@ -0,0 +1 @@
+WS2812_DRIVER = spi
```
Comment (zvecr) [inline keyboards/plywrks/ply8x/rules.mk:1]:
```suggestion
# This file intentionally left blank
```

File: keyboards/plywrks/ply8x/info.json [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x7905",
+        "vid": "0x706C"
+    },
+    "rgb_matrix": {
+        "driver": "WS2812",
+        "layout": [
+            {"flags": 8, "matrix": [3, 0], "x":2, "y":27},
+            {"flags": 8, "matrix": [0,15], "x":211, "y":0}
```
Comment (zvecr) [inline keyboards/plywrks/ply8x/info.json:31]: > #define RGB_MATRIX_LED_COUNT 6

Only 2 entries here is invalid with the previous config

File: keyboards/plywrks/ply8x/info.json [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        "rows": ["B2", "B1", "B0", "A7", "A10", "A2"],
+        "cols": ["A9", "A8", "B12", "B11", "B10", "A6", "A5", "A4", "A15", "C13", "C14", "B9", "B7", "B6", "B5", "B4", "B3"]
+    },
+    "processor": "STM32F072",
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x7905",
+        "vid": "0x706C"
+    },
```
Comment (zvecr) [inline keyboards/plywrks/ply8x/info.json:26]:
```suggestion
    },
    "ws2812": {
        "driver": "spi"
    }.
```

File: keyboards/plywrks/ply8x/keymaps/via/keymap.c [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        _______, _______, _______,                            _______,                            _______, _______, _______, _______,          _______, _______, _______
+    ),
+    [3] = LAYOUT_all(
+        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,                   _______,
+        _______, _______, _______,                            _______,                            _______, _______, _______, _______,          _______, _______, _______
+    ),
```
Comment (zvecr) [inline keyboards/plywrks/ply8x/keymaps/via/keymap.c:38]:
```suggestion
    )
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/plywrks/ply8x/info.json (hunk 2 above)
Comment (ramonimbao) [reply keyboards/plywrks/ply8x/info.json]: Oops, sorry about that. Forgot to add the new LEDs after updating the number of LEDs.

File: keyboards/plywrks/ply8x/config.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+/* RGB */
+#ifdef RGB_MATRIX_ENABLE
+#define WS2812_SPI SPID2
+#define WS2812_SPI_MOSI_PAL_MODE 0
+#define WS2812_SPI_SCK_PIN B13
+#define WS2812_SPI_SCK_PAL_MODE 0
+
+#define WS2812_EXTERNAL_PULLUP
+#define WS2812_DI_PIN B15
```
Comment (waffle87) [inline keyboards/plywrks/ply8x/config.h:14]:
```suggestion
```

File: keyboards/plywrks/ply8x/config.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+/* disable debug print */
+//#define NO_DEBUG
+
+/* disable print */
+//#define NO_PRINT
+
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
```
Comment (waffle87) [inline keyboards/plywrks/ply8x/config.h:79]:
```suggestion
```

File: keyboards/plywrks/ply8x/info.json [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    "ws2812": {
```
Comment (waffle87) [inline keyboards/plywrks/ply8x/info.json:27]:
```suggestion
    "ws2812": {
        "pin": "B15",
```

File: keyboards/plywrks/ply8x/keymaps/via/keymap.c [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,          KC_LEFT, KC_DOWN, KC_RGHT
+    ),
+    [1] = LAYOUT_all(
+        _______,          _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______,
+        _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,                   _______,
+        _______, _______, _______,                            _______,                            _______, _______, _______, _______,          _______, _______, _______
```
Comment (waffle87) [inline keyboards/plywrks/ply8x/keymaps/via/keymap.c:21]:
```suggestion
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/plywrks/ply8x/config.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_WIDE
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTIWIDE
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_CROSS
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTICROSS
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_NEXUS
+#define ENABLE_RGB_MATRIX_SOLID_REACTIVE_MULTINEXUS
+#define ENABLE_RGB_MATRIX_SPLASH
+#define ENABLE_RGB_MATRIX_MULTISPLASH
+#define ENABLE_RGB_MATRIX_SOLID_SPLASH
+#define ENABLE_RGB_MATRIX_SOLID_MULTISPLASH
```
Comment (dunk2k) [inline keyboards/plywrks/ply8x/config.h:61]:
```suggestion
#define WS2812_EXTERNAL_PULLUP
#define RGB_MATRIX_LED_COUNT 6
#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
#define RGB_MATRIX_KEYPRESSES
```

File: keyboards/plywrks/ply8x/info.json [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        "pin": "B15",
+        "driver": "spi"
+    },
+    "rgb_matrix": {
```
Comment (dunk2k) [inline keyboards/plywrks/ply8x/info.json:31]:
```suggestion
    "rgb_matrix": {
        "animations": {
            "alphas_mods": true,
            "gradient_up_down": true,
            "gradient_left_right": true,
            "breathing": true,
            "band_sat": true,
            "band_val": true,
            "band_pinwheel_sat": true,
            "band_pinwheel_val": true,
            "band_spiral_sat": true,
            "band_spiral_val": true,
            "cycle_all": true,
            "cycle_left_right": true,
            "cycle_up_down": true,
            "rainbow_moving_chevron": true,
            "cycle_out_in": true,
            "cycle_out_in_dual": true,
            "cycle_pinwheel": true,
            "cycle_spiral": true,
            "dual_beacon": true,
            "rainbow_beacon": true,
            "rainbow_pinwheels": true,
            "raindrops": true,
            "jellybean_raindrops": true,
            "hue_breathing": true,
            "hue_pendulum": true,
            "hue_wave": true,
            "pixel_rain": true,
            "pixel_flow": true,
            "pixel_fractal": true,
            "typing_heatmap": true,
            "digital_rain": true,
            "solid_reactive_simple": true,
            "solid_reactive": true,
            "solid_reactive_wide": true,
            "solid_reactive_multiwide": true,
            "solid_reactive_cross": true,
            "solid_reactive_multicross": true,
            "solid_reactive_nexus": true,
            "solid_reactive_multinexus": true,
            "splash": true,
            "multisplash": true,
            "solid_splash": true,
            "solid_multisplash": true
        },
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

**Comment by dunk2k** [comment]
@ramonimbao 
Looking at the image of the PCB (rear facing), it appears there are accommodations for additional physical layouts (split_backspace, split right shift, iso?). Would you be able to add all the possible physical layouts as additional layouts in `info.json` and create a keymap for each?

**Comment by ramonimbao** [comment]
> Would you be able to add physical layouts as additional layouts in info.json and create a keymap for each?

Is this really necessary? The boards will definitely ship with the via keymap pre-flashed.

**Comment by dunk2k** [comment]
> Is this really necessary? The boards will definitely ship with the via keymap pre-flashed.

necessary, I'd say, is dependent if a user/buyer could actually (i.e. both electronically and physically) have switches arranged differently to the default - PR #21201 is a good example.

**Comment by ramonimbao** [comment]
Yeah, but that can also be done with VIA by selecting which layout options to enable/disable.

**Comment by fauxpark** [comment]
And what about users who don't want to or cannot use VIA?

**Comment by dunk2k** [comment]
> Yeah, but that can also be done with VIA by selecting which layout options to enable/disable.

Which is great for VIA users, but no help for QMK users.

**Comment by ramonimbao** [comment]
Hmm alright. I'll add the additional layouts.

File: keyboards/plywrks/ply8x/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    )
+};
+
+#ifdef RGB_MATRIX_ENABLE
+bool rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max) {
+    if (!((host_keyboard_led_state().scroll_lock))) {
+        RGB_MATRIX_INDICATOR_SET_COLOR(1, 0, 0, 0);
+    }
+
+    if (!((host_keyboard_led_state().caps_lock))) {
```
Comment (fauxpark) [inline keyboards/plywrks/ply8x/keymaps/default/keymap.c:40]:
```suggestion
    led_t led_state = host_keyboard_led_state();
    if (!led_state..scroll_lock) {
        RGB_MATRIX_INDICATOR_SET_COLOR(1, 0, 0, 0);
    }

    if (!led_state.caps_lock) {
```

File: keyboards/plywrks/ply8x/keymaps/via/keymap.c [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+    )
+};
+
+#ifdef RGB_MATRIX_ENABLE
+bool rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max) {
+    if (!((host_keyboard_led_state().scroll_lock))) {
+        RGB_MATRIX_INDICATOR_SET_COLOR(1, 0, 0, 0);
+    }
+
+    if (!((host_keyboard_led_state().caps_lock))) {
```
Comment (fauxpark) [inline keyboards/plywrks/ply8x/keymaps/via/keymap.c:23]:
```suggestion
    led_t led_state = host_keyboard_led_state();
    if (!led_state.scroll_lock) {
        RGB_MATRIX_INDICATOR_SET_COLOR(1, 0, 0, 0);
    }

    if (!led_state.caps_lock) {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/plywrks/ply8x/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,
+        KC_LSFT,          KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,             KC_UP,
+        KC_LCTL, KC_LGUI, KC_LALT,                            KC_SPC,                             KC_RALT, KC_RGUI, KC_APP,  KC_RCTL,    KC_LEFT, KC_DOWN, KC_RGHT
+    )
+};
+
+#ifdef RGB_MATRIX_ENABLE
+bool rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max) {
+    led_t led_state = host_keyboard_led_state();
+    if (!led_state..scroll_lock) {
```
Comment (dunk2k) [inline keyboards/plywrks/ply8x/keymaps/default/keymap.c:37]: fix CI Build failure
```suggestion
    if (!led_state.scroll_lock) {
```

**Review by dunk2k — CHANGES_REQUESTED** [review]

**Review by dunk2k — APPROVED** [review]

File: keyboards/plywrks/ply8x/readme.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+* Keyboard Maintainer: [Ramon Imbao](https://github.com/ramonimbao)
+* Hardware Supported: STM32F072
+
+Make example for this keyboard (after setting up your build environment):
+
+    make plywrks/ply8x:via
+
+Flashing example for this keyboard:
+
+    make plywrks/ply8x:via:flash
```
Comment (fauxpark) [inline keyboards/plywrks/ply8x/readme.md:16]:
```suggestion
    make plywrks/ply8x:default

Flashing example for this keyboard:

    make plywrks/ply8x:default:flash
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by waffle87 — APPROVED** [review]

File: keyboards/plywrks/ply8x/config.h [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+#ifdef RGB_MATRIX_ENABLE
+#define WS2812_SPI SPID2
+#define WS2812_SPI_MOSI_PAL_MODE 0
+#define WS2812_SPI_SCK_PIN B13
+#define WS2812_SPI_SCK_PAL_MODE 0
+#define WS2812_EXTERNAL_PULLUP
+#define RGB_MATRIX_LED_COUNT 6
+#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
+#define RGB_MATRIX_KEYPRESSES
+#endif
```
Comment (fauxpark) [inline keyboards/plywrks/ply8x/config.h:16]:
```suggestion
#define WS2812_SPI SPID2
#define WS2812_SPI_MOSI_PAL_MODE 0
#define WS2812_SPI_SCK_PIN B13
#define WS2812_SPI_SCK_PAL_MODE 0
#define WS2812_EXTERNAL_PULLUP
#define RGB_MATRIX_LED_COUNT 6
#define RGB_MATRIX_FRAMEBUFFER_EFFECTS
#define RGB_MATRIX_KEYPRESSES
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Comment by ramonimbao** [comment]
Thanks everyone 🙏

**Pull request merged**